### DISCUSSION TITLE
Standalone WinUI C++ or with XAML anyway

### DISCUSSION THREAD

**Discussion by TheHellTower** [microsoft/WindowsAppSDK#3359] [Q&A] [answered]
Hello,

I'm interested in C++ GUI but the default option to create C++ GUI in Visual Studio is really worse (longer to update and implement new stuff) while WinUI for example use XAML so I seen you can completely export as normal executable but it does give a lot of DLLs and Folders next to it, I wonder if it's possible to achieve the same but with a standalone executable 
instead ?

Because the really minimal I can go to is [this](https://i.imgur.com/hKYl6LV.png)..

Regards.

**Comment by DarranRowe** [comment, marked as answer]
To put it simply, no.
The Windows App SDK isn't a standard part of Windows, this means you need the redistributable to run. The Windows App SDK doesn't provide static libraries, this means that you can't combine everything into a single executable.
This is actually why relying on the Windows App SDK framework package is the preferred option right now.

**Reply by TheHellTower** [reply]
> To put it simply, no. The Windows App SDK isn't a standard part of Windows, this means you need the redistributable to run. The Windows App SDK doesn't provide static libraries, this means that you can't combine everything into a single executable. This is actually why relying on the Windows App SDK framework package is the preferred option right now.

Hello,

Thanks for your answer, but is there any way to make GUI using XAML or anything else from Visual Studio ? I don't want to rely on QT that add a lot of DLLs too.

Regards.

**Reply by DarranRowe** [reply]
Unfortunately, you will have to accept a trade off if you want to use Xaml. As mentioned already, there is no static libraries for Xaml, so if you want to avoid any extra DLLs then you will have to use inbox libraries. For C++, this means UWP Xaml through Xaml Islands.
The trade off here is that Xaml Islands in a desktop application is not the same as a regular Xaml application, and you have to do a fair amount of extra work to get things hooked up properly.
There is a lot more to say on this subject, but if your goal is to make a single executable download then it is pretty much impossible if you want to use Xaml.

**Reply by TheHellTower** [reply]
> Unfortunately, you will have to accept a trade off if you want to use Xaml. As mentioned already, there is no static libraries for Xaml, so if you want to avoid any extra DLLs then you will have to use inbox libraries. For C++, this means UWP Xaml through Xaml Islands. The trade off here is that Xaml Islands in a desktop application is not the same as a regular Xaml application, and you have to do a fair amount of extra work to get things hooked up properly. There is a lot more to say on this subject, but if your goal is to make a single executable download then it is pretty much impossible if you want to use Xaml.

Ok then I will stay with the minimal result from the screenshot thanks for everything !

**Comment by driver1998** [comment]
If you are willing to use the older builtin XAML layer with XAML islands, you can get pretty small with C++, but still not single file.

**Reply by TheHellTower** [reply]
> If you are willing to use the older builtin XAML layer with XAML islands, you can get pretty small with C++, but still not single file.

Hello,

If it can permit to lower more the size and doesn't limit me too much in possibilities it doesn't matter I can go with it I will just need to know how,

Regards.

**Reply by driver1998** [reply]
You can start with the official docs: https://learn.microsoft.com/en-us/windows/apps/desktop/modernize/xaml-islands.
It will be a little bit daunting since XAML islands is more towards mixing XAML with existing UI, not building a whole app with it.

You can also take a look at Mile.Xaml, which is a third-party minimal scaffold for XAML-island-only apps.
https://github.com/ProjectMile/Mile.Xaml

Just as an example, this is a demo by the author of Mile.Xaml @MouriNaruto, which is about 1MB all together.
![20230129_033327.jpg](https://user-images.githubusercontent.com/22699485/215292983-e8731aa3-3f2d-436c-ae3f-e465c5a1eced.jpg)

**Reply by TheHellTower** [reply]
> You can start with the official docs: https://learn.microsoft.com/en-us/windows/apps/desktop/modernize/xaml-islands. It will be a little bit daunting since XAML islands is more towards mixing XAML with existing UI, not building a whole app with it.
> 
> You can also take a look at Mile.Xaml, which is a third-party minimal scaffold for XAML-island-only apps. https://github.com/ProjectMile/Mile.Xaml
> 
> Just as an example, this is a demo by the author of Mile.Xaml @MouriNaruto, which is about 1MB all together. ![20230129_033327.jpg](https://user-images.githubusercontent.com/22699485/215292983-e8731aa3-3f2d-436c-ae3f-e465c5a1eced.jpg)

It will be perfect thanks for your help it's already a lot better !

**Reply by TheHellTower** [reply]
Perfect !
![](https://i.imgur.com/VfIwGe4.png)
